### PR TITLE
Packaging

### LINKED ISSUE #1687: Publish rpm and deb packages
We removed support for rpm and deb packages in https://github.com/grafana/mimir/pull/1138 but there's some people in the community that finds them useful and we should re-introduce them.

Thoughts:
- Previously, they were configured with `/etc/cortex/single-process-config.yaml` by default. I would suggest to not bind them to any specific config. By default, Mimir should run with the default configuration. It should be easy to specify a config file path tho.
- We need to discuss at Grafana Labs where these packages should be published.

### PR DESCRIPTION
**Pull request by wilfriedroset**
#### What this PR does

Reintroducing packaging for both debian based and redhat based using FPM.

#### Which issue(s) this PR fixes or relates to

Fixes #1687

#### Checklist

- [ ] Tests updated
- [x] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by wilfriedroset** [comment]
I did not managed to get the target `test-packages` working as running systemd in docker in trivial.
Would a maintainer point me in the right direction to get everything in order?

I can attached the packages if necessary or they can be built using `make packages` as explained in `RELEASE.md`

**Comment by wilfriedroset** [comment]
rebased on latest main commits

**Comment by colega** [comment]
Hi, sorry for the delay. It doesn't seem that we're going to be able to include this in the 2.2 release (because of our delay, we're running too late for that now), but we're definitely going to try to make it part of the 2.3.

File: packaging/fpm/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# SPDX-License-Identifier: AGPL-3.0-only
+# Provenance-includes-location: https://github.com/grafana/mimir/packaging/fpm/Dockerfile
+# Provenance-includes-license: Apache-2.0
+# Provenance-includes-copyright: Grafana Labs.
+
+FROM alpine:3.8
```
Comment (pstibrany) [inline packaging/fpm/Dockerfile:6]: To be consistent with our other Dockerfiles.

```suggestion
FROM alpine:3.16.0
```

File: packaging/deb/debian-systemd/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# SPDX-License-Identifier: AGPL-3.0-only
+# Provenance-includes-location: https://github.com/grafana/mimir/packaging/deb/debian-systemd/Dockerfile
```
Comment (pstibrany) [inline packaging/deb/debian-systemd/Dockerfile:2]:
```suggestion
# Provenance-includes-location: https://github.com/cortexproject/cortex/blob/master/packaging/deb/debian-systemd/Dockerfile
```

(This should be done for all Provenance-includes-location in the PR.)

File: packaging/deb/debian-systemd/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+# Provenance-includes-license: Apache-2.0
+# Provenance-includes-copyright: Grafana Labs.
```
Comment (pstibrany) [inline packaging/deb/debian-systemd/Dockerfile:4]:
```suggestion
# Provenance-includes-copyright: The Cortex Authors.
```

(to be applied for all `Provenance-includes-copyright` headers in this PR)

File: packaging/deb/debian-systemd/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+FROM debian:10
```
Comment (pstibrany) [inline packaging/deb/debian-systemd/Dockerfile:6]: Latest is now 11 (bullseye), I think we should tests against that.

**Review by pstibrany — COMMENTED** [review]
Thank you. I've left some minor comments, but overall this looks fine.

I've noticed that License and Provenance headers are missing in many scripts (eg. post/prerun) and config files (mimir.service). I wonder if we should add them everywhere. @pracucci WDYT?

**Comment by wilfriedroset** [comment]
> Thank you. I've left some minor comments, but overall this looks fine.
> 
> I've noticed that License and Provenance headers are missing in many scripts (eg. post/prerun) and config files (mimir.service). I wonder if we should add them everywhere. @pracucci WDYT?

I agree about the headers which I've added. I've also taken into account all your comments and rebase on main to solve conflits.

File: Makefile [hunk 5]
```diff
@@ -523,9 +529,83 @@ check-jsonnet-tests: build-jsonnet-tests
 check-tsdb-blocks-storage-s3-docker-compose-yaml:
 	cd development/tsdb-blocks-storage-s3 && make check
 
+# Generate packages for a Mimir release.
+FPM_OPTS := fpm -s dir -v $(VERSION) -n mimir -f \
+	--license "AGPL 3.0" \
+	--url "https://github.com/grafana/mimir" \
```
Comment (pstibrany) [inline Makefile:535]: What does this URL refer to? In service files, we use `https://grafana.com/oss/mimir/` instead.

File: Makefile [hunk 6]
```diff
@@ -523,9 +529,83 @@ check-jsonnet-tests: build-jsonnet-tests
...
+	--description "Grafana Mimir provides horizontally scalable, highly available, multi-tenant, long-term storage for Prometheus." \
+	--maintainer "contact@grafana.com" \
+	--vendor "Grafana"
```
Comment (pstibrany) [inline Makefile:538]:
```suggestion
	--vendor "Grafana Labs"
```

**Review by pstibrany — APPROVED** [review]
LGTM, with some minor nits.

Thank you for working on this!

File: Makefile (hunk 5 above)
Comment (wilfriedroset) [reply Makefile]: This URL ends up in the documentation link of the service file
```
root@f4d35967e37d:/mnt# cat /lib/systemd/system/mimir.service
# SPDX-License-Identifier: AGPL-3.0-only
# Provenance-includes-location: https://github.com/cortexproject/cortex/blob/master/packaging/deb/systemd/cortex.service
# Provenance-includes-license: Apache-2.0
# Provenance-includes-copyright: The Cortex Authors.

[Unit]
Description=Horizontally scalable, highly available, multi-tenant, long term Prometheus.
Documentation=https://grafana.com/oss/mimir/
Wants=network-online.target
After=network-online.target

[Service]
Restart=always
User=mimir
EnvironmentFile=/etc/default/mimir
ExecStart=/usr/local/bin/mimir --log.level $LOG_LEVEL $CUSTOM_ARGS
ExecReload=/bin/kill -HUP $MAINPID
TimeoutStopSec=20s
SendSIGKILL=no
WorkingDirectory=/var/lib/mimir

[Install]
WantedBy=multi-user.target
```

**Comment by wilfriedroset** [comment]
Rebase to address conflict 😅

**Review by treid314 — APPROVED** [review]
Looks great! Thank you! I'll add a CHANGELOG message as part of the release process.

**Pull request merged**